### ISSUE TITLE
Implicit preconditions - incorrect number of parameters

### ISSUE THREAD

**Issue by joey-coleman** [symphonytool/symphony#110]
> The following bug was [originally reported](https://sourceforge.net/p/compassresearch/bugs/110/) on Sourceforge by [*anonymous](https://sourceforge.net/u/*anonymous/profile/), 2013-10-09 10:19:08.194000:

When quoting a precondition expression of a defined implicit operation, the incorrect number of parameters is required. For example, if we have an operation o1:

o1(n5:nat, n6: nat) r: nat
pre n5 > 66
post r = n5 \* n6

I would expect to be able to use the precondition in an expression as follows:

pre_o1(n7,n8)

However I get an error: "Too few arguments". Adding an extra parameter removes the error. 

All other precondition expressions (explicit/implicit functions and explicit operations) take the same number of parameters as the owning function/operation.

My test CML model is below:

process proc = 
begin
  operations

```
o1 : nat * nat ==> nat
o1(n1, n2) ==  return n1* n2
pre n1>n2
post RESULT > n1

o2 : nat * nat ==> nat
o2(n3, n4) == return n3 * n4
pre pre_o1(n3,n4) and n4 > 66


o3(n5:nat, n6: nat) r: nat
pre n5 > 66
post r = n5 * n6


o4(n7:nat, n8: nat) r: nat
pre pre_o3(n7,n8) and n8 > 66
post r = 55

@ Skip
```

end

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/110/#937f) by [joeycoleman](https://sourceforge.net/u/joeycoleman/profile/), 2013-10-09 10:33:19.131000:
- **assigned_to**: Claus B. Nielsen

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/110/#228c) by [joeycoleman](https://sourceforge.net/u/joeycoleman/profile/), 2013-10-09 10:33:19.461000:

I'm guessing that the generated function is sticking the operation's return variable into the pre_op()'s parameter list; we should check if it happens on operations with no return.

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/110/#ef8a) by [clausbn](https://sourceforge.net/u/clausbn/profile/), 2013-10-11 17:16:25.921000:

The source that builds the invariant specifically looks for a ImplicitCmlOperationDefinition and adds the expected result of the operation to the parameter list. 

Am unsure why this has been done.

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/110/#f37c) by [clausbn](https://sourceforge.net/u/clausbn/profile/), 2013-10-11 17:16:48.989000:
- **status**: open --> accepted

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/110/#6bd1) by [clausbn](https://sourceforge.net/u/clausbn/profile/), 2013-10-16 16:07:59.009000:

The typechecker now allows the precondition to be called without the result. 

Fixed by [c724cc5a786ea1d8680c2d0940b7c5cd4a6df8f9]

**Comment by joey-coleman** [comment]
> [Comment](https://sourceforge.net/p/compassresearch/bugs/110/#493d) by [clausbn](https://sourceforge.net/u/clausbn/profile/), 2013-10-16 16:08:10.407000:
- **status**: accepted --> closed
